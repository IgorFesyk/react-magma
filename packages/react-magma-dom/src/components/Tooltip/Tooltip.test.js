import React from 'react';
import { axe } from 'jest-axe';
import { Tooltip } from '.';

import { render, fireEvent } from 'react-testing-library';

const CONTENT_TEXT = 'Test Content';
const TRIGGER_ELEMENT = <button>Test trigger</button>;

describe('Tooltip', () => {
  it('should render the tooltip component, positioned bottom by default', () => {
    const { container, getByText } = render(
      <Tooltip content={CONTENT_TEXT} trigger={TRIGGER_ELEMENT} />
    );
    const tooltip = getByText('Test Content');
    const tooltipTrigger = container.firstChild;

    expect(container).toBeInTheDocument();

    expect(tooltipTrigger).toBeInTheDocument();
    expect(tooltipTrigger).toHaveStyleRule('position', 'relative');

    expect(tooltip).toBeInTheDocument();
    expect(tooltip).toHaveStyleRule('position', 'absolute');
    expect(tooltip).toHaveStyleRule('top', '100%');
  });

  it('should render the tooltip component with the correct styles when positioned left', () => {
    const { getByText } = render(
      <Tooltip
        content={CONTENT_TEXT}
        position="left"
        trigger={TRIGGER_ELEMENT}
      />
    );
    const tooltip = getByText('Test Content');

    expect(tooltip).toHaveStyleRule('right', '100%');
  });

  it('should render the tooltip component with the correct styles when positioned right', () => {
    const { getByText } = render(
      <Tooltip
        content={CONTENT_TEXT}
        position="right"
        trigger={TRIGGER_ELEMENT}
      />
    );
    const tooltip = getByText('Test Content');

    expect(tooltip).toHaveStyleRule('left', '100%');
  });

  it('should render the tooltip component with the correct styles when positioned top', () => {
    const { getByText } = render(
      <Tooltip
        content={CONTENT_TEXT}
        position="top"
        trigger={TRIGGER_ELEMENT}
      />
    );
    const tooltip = getByText('Test Content');

    expect(tooltip).toHaveStyleRule('bottom', '100%');
  });

  it('should show the tooltip on focus and hide it on blur', () => {
    const { getByText } = render(
      <Tooltip content={CONTENT_TEXT} trigger={TRIGGER_ELEMENT} />
    );
    const trigger = getByText('Test trigger');
    const tooltip = getByText('Test Content');
    expect(tooltip).toHaveStyleRule('display', 'none');

    fireEvent.focus(trigger);

    expect(tooltip).toHaveStyleRule('display', 'block');
    fireEvent.blur(trigger);

    expect(tooltip).toHaveStyleRule('display', 'none');
  });

  it('should show the tooltip on mouseenter and hide it on mouseleave', () => {
    const { getByText } = render(
      <Tooltip content={CONTENT_TEXT} trigger={TRIGGER_ELEMENT} />
    );
    const trigger = getByText('Test trigger');
    const tooltip = getByText('Test Content');
    expect(tooltip).toHaveStyleRule('display', 'none');

    fireEvent.mouseEnter(trigger);

    expect(tooltip).toHaveStyleRule('display', 'block');
    fireEvent.mouseLeave(trigger);

    expect(tooltip).toHaveStyleRule('display', 'none');
  });

  it('should hide the tooltip when the escape key is pressed', () => {
    const { getByText } = render(
      <Tooltip content={CONTENT_TEXT} trigger={TRIGGER_ELEMENT} />
    );
    const trigger = getByText('Test trigger');
    const tooltip = getByText('Test Content');
    expect(tooltip).toHaveStyleRule('display', 'none');

    fireEvent.focus(trigger);

    expect(tooltip).toHaveStyleRule('display', 'block');

    fireEvent.keyDown(trigger, {
      key: 'ArrowDown',
      code: 40
    });

    expect(tooltip).toHaveStyleRule('display', 'block');

    fireEvent.keyDown(trigger, {
      key: 'Escape',
      keyCode: 27
    });

    expect(tooltip).toHaveStyleRule('display', 'none');
  });

  it('Does not violate accessibility standards', () => {
    const { container } = render(
      <Tooltip content={CONTENT_TEXT} trigger={TRIGGER_ELEMENT} />
    );

    return axe(container.innerHTML).then(result => {
      return expect(result).toHaveNoViolations();
    });
  });
});
