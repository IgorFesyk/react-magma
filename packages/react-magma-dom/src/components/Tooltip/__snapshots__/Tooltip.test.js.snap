// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should render the tooltip component, positioned top by default 1`] = `
.emotion-1 {
  display: none;
  font-size: 12px;
  line-height: 1.3em;
  font-weight: 600;
  position: absolute;
  text-align: center;
  width: 300px;
  z-index: 999;
  bottom: 100%;
  margin-bottom: 10px;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.emotion-0 {
  background: #3F3F3F;
  border-radius: 3px;
  color: #FFFFFF;
  display: inline-block;
  padding: 7px 10px;
  position: relative;
}

.emotion-0:before,
.emotion-0:after {
  border-left-color: transparent;
  border-right-color: transparent;
  border-top-color: #3F3F3F;
  border-bottom-color: #3F3F3F;
  border-style: solid;
  content: '';
  height: 0;
  left: 50%;
  position: absolute;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  width: 0;
}

.emotion-0:after {
  bottom: -5px;
  top: auto;
  border-width: 5px 5px 0 5px;
}

.emotion-0:before {
  bottom: -7px;
  top: auto;
  border-width: 7px 7px 0 7px;
}

<div
  aria-hidden="true"
  class="emotion-1"
  id="tooltipID"
  role="tooltip"
>
  <div
    class="emotion-0"
  >
    Test Content
  </div>
</div>
`;
